### PR TITLE
feat(core): store cause of group-resource assignment failure

### PR DESCRIPTION
**Pull request by Johaney-s**
* The error message is now saved to DB when assignment fails
* When the assignment state is switched to ACTIVE or INACTIVE, previous cause of failure is removed
* The cause can only be retrieved for FAILED assignments

### CODE REVIEW

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ResourcesManagerBlImpl.java [hunk 1]
```diff
@@ -1175,6 +1181,19 @@ public void deactivateGroupResourceAssignment(PerunSession sess, Group group, Re
...
+		// reset assignment failure cause
+		getResourcesManagerImpl().setFailedGroupResourceAssignmentCause(sess, group, resource, null);
+	}
+
+	@Override
+	public String getFailedGroupResourceAssignmentCause(PerunSession sess, Group group, Resource resource) throws GroupNotDefinedOnResourceException, GroupResourceStatusException {
+		GroupResourceStatus status = getResourcesManagerImpl().getGroupResourceStatus(sess, group, resource);
+		if (status != GroupResourceStatus.FAILED) {
+			throw new GroupResourceStatusException("Cannot read assignment failure cause for not FAILED assignment.");
+		}
```
Comment (Vojtech-Sassmann) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ResourcesManagerBlImpl.java:1194]: In such scenarios, it is better to just return `null` value (or `Optional<String>` if you want to do it the best way). If you throw an exception, you limit the usage of such a method (you force the method users to handle the exception even if they first check that the status is failing for example).

**Review by Vojtech-Sassmann — CHANGES_REQUESTED** [review]
I'm not sure if we don't want to put the cause directly to objects like `AssignedGroup` and `AssignedResource`? I don't think it is necessary, but it could make life easier for GUI development. What do you think?

**Comment by Johaney-s** [comment]
> I'm not sure if we don't want to put the cause directly to objects like `AssignedGroup` and `AssignedResource`? I don't think it is necessary, but it could make life easier for GUI development. What do you think?

Well, what I was thinking is the cause would be retrieved only in case user clicks the failed status icon, so not very often. But I'll ask GUI people what they prefer and change it accordingly.

**Comment by Vojtech-Sassmann** [comment]
> > I'm not sure if we don't want to put the cause directly to objects like `AssignedGroup` and `AssignedResource`? I don't think it is necessary, but it could make life easier for GUI development. What do you think?
> 
> Well, what I was thinking is the cause would be retrieved only in case user clicks the failed status icon, so not very often. But I'll ask GUI people what they prefer and change it accordingly.

Yeah, I also think this way it is sufficient. If they will agree with you, you can keep it this way.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ResourcesManagerBlImpl.java (hunk 1 above)
Comment (Johaney-s) [reply perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ResourcesManagerBlImpl.java]: Added `failureCause` to `AssignedResource`, `AssignedGroup` and `GroupResourceAssignment`, replaced throwing `GroupResourceStatusException` for returning `null`.

File: perun-base/src/main/resources/perun-roles.yml [hunk 2]
```diff
@@ -3916,6 +3916,17 @@ perun_policies:
...
+    policy_roles:
+      - RESOURCEADMIN: Resource
+      - VOADMIN: Vo
+      - RESOURCESELFSERVICE: Resource
+        GROUPADMIN: Group
+      - TRUSTEDFACILITYADMIN: Vo
+        FACILITYADMIN: Facility
+    include_policies:
+      - default_policy
+
```
Comment (Vojtech-Sassmann) [inline perun-base/src/main/resources/perun-roles.yml:3929]: This can be probably removed.

**Review by Vojtech-Sassmann — CHANGES_REQUESTED** [review]

File: perun-base/src/main/resources/perun-roles.yml (hunk 2 above)
Comment (Johaney-s) [reply perun-base/src/main/resources/perun-roles.yml]: Removed

File: perun-core/src/main/resources/postgresChangelog.txt [hunk 3]
```diff
@@ -6,6 +6,10 @@
 -- Directly under version number should be version commands. They will be executed in the order they are written here.
 -- Comments are prefixed with -- and can be written only between version blocks, that means not in the lines with commands. They have to be at the start of the line.
 
+3.1.84
+ALTER TABLE groups_resources ADD COLUMN failure_cause default null;
```
Comment (cuadradek) [inline perun-core/src/main/resources/postgresChangelog.txt:10]:
```suggestion
ALTER TABLE groups_resources ADD COLUMN failure_cause varchar default null;
```

**Review by cuadradek — CHANGES_REQUESTED** [review]

File: perun-core/src/main/resources/postgresChangelog.txt (hunk 3 above)
Comment (Johaney-s) [reply perun-core/src/main/resources/postgresChangelog.txt]: thanks, added

**Review by cuadradek — APPROVED** [review]

**Pull request merged**